src/auslib/services/releases.py

```diff
@@ -300,16 +300,52 @@ def get_releases(trans):
 
 
 def get_release(name, trans):
+    def get_base_row():
+        row = dbo.releases_json.select(where={"name": name}, transaction=trans)
+        if row:
+            return row[0]
+
+        return None
+
+    def get_base_data_version():
+        row = dbo.releases_json.select(where={"name": name}, columns=[dbo.releases_json.data_version], transaction=trans)
+        if row:
+            return row[0]["data_version"]
+
+        return None
+
+    def get_asset_rows():
+        return dbo.release_assets.select(where={"name": name}, transaction=trans) or []
+
+    def get_asset_data_versions():
+        return dbo.releases_json.select(where={"name": name}, columns=[dbo.releases_json.data_version], transaction=trans) or []
+
+    # Get all of the base and asset information, potentially from a cache
+    base_row = cache.get("releases", name, get_base_row)
+    base_data_version = cache.get("releases_data_version", name, get_base_data_version)
+    asset_rows = cache.get("release_assets", name, get_asset_rows)
+    asset_data_versions = cache.get("release_assets_data_versions", name, get_asset_data_versions)
+
     data_versions = infinite_defaultdict()
     sc_data_versions = infinite_defaultdict()
     base_blob = {}
     sc_blob = {}
-    base_row = dbo.releases_json.select(where={"name": name}, transaction=trans)
     if base_row:
-        base_blob = base_row[0]["data"]
-        data_versions["."] = base_row[0]["data_version"]
-
-    for asset in dbo.release_assets.select(where={"name": name}, transaction=trans):
+        # base_data_version is cached for a shorter period of time than the overall row
+        # because it's cheap to retrieve. if the cached row's data_version is older than
+        # the cached data_version we will forcibly update it to make sure we minimize
+        # the time we're serving old release data
+        if base_row["data_version"] < base_data_version:
+            base_row = get_base_row()
+        base_blob = base_row["data"]
+        data_versions["."] = base_row["data_version"]
+
+    # same thing here for the assets -- if any of the full asset data versions
+    # do not match the cached asset data versions, we forcibly update
+    if [r["data_version"] for r in asset_rows] != asset_data_versions:
+        asset_rows = get_asset_rows()
```

Comment:
This is a tough question to answer accurately in local development or an ad-hoc GCP environment because of load/instance size/etc. I was debating about which way to go on this myself, and I _think_ this is going to end up being better because the assets are almost always unchanging after we build a Release - so the likelyhood of any one of the locales changing is near zero. In the real world, this query is most likely to run once per Release per hour per webhead per process/thread - so the performance impact should be negligible. This should be true for update verify as well, since it blocks on `post-beetmover-dummy` (although I can imagine a world where it only blocks on specific upstreams it needs).

If we retrieved and cached per locale, we'd end up running ~500 queries per Release (per hour, etc.), and I think the overhead ends up making it slower overall (although likely still negligible).

Again, this is all slightly speculative because it's hard to be certain what production performance will be but it's something I will be looking at closely the first time we ship something with one of these Releases.

Reply:
OK, that makes lots of sense. It may be different while we're building a new nightly in the current system, but I haven't kept up with the nightly plan so perhaps it'll go away there too. Lets stick with your approach here.

Reply:
That's a very good point actually. The plan right now is to continue using the `latest` blobs (although a new one, in the new tables when we migrate) as part of this transition. I believe @Callek is planning to get them in line with Releases in that we'd only ship after everything is built - but that work will be happening after this is done.

This means, of course, that we're likely to see a lot more cache invalidation for nightlies if/when this lands as is. If it becomes an issue we can make a change to the caching.